### ISSUE TITLE
Using zig cc to compile acton leads to illegal instruction

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#989]
Using `zig cc` to compile `builtin/hash.c` leads to `SIGILL`.
```c
long $pointer_hash($WORD p) {
    long x;
    long y = (long)p;
    // bottom 3 or 4 bits are likely to be 0; rotate y by 4 to avoid
    //  excessive hash collisions for dicts and sets 
    y = (y >> 4) | (y << 60);
    x = (long)y;
    if (x == -1)
        x = -2;
    return x;
}
```

Specifically, it's the bit shifting line that is causing this which we can reasily reproduce with this test program that also outputs a binary view to ease understanding:
```c
#include <stdio.h>

// Assumes little endian
void printBits(size_t const size, void const * const ptr)
{
    unsigned char *b = (unsigned char*) ptr;
    unsigned char byte;
    int i, j;

    for (i = size-1; i >= 0; i--) {
        for (j = 7; j >= 0; j--) {
            byte = (b[i] >> j) & 1;
            printf("%u", byte);
        }
    }
    puts("");
}

int main(int argc, char **argv) {
        long y = 1;
        y = (y >> 4) | (y << 60);
        long a = y;

        printf("a: %lu\n", a);
        printBits(sizeof(a), &a);
        printf("\n");
        return 0;
}
```
I can compile and run this with gcc & zig cc like so:
```
kll@Boxy:~/terastream/acton/testzig$ gcc test.c -o outgcc && zig cc test.c -o outzig && ./outgcc && ./outzig
a: 1152921504606846976
0001000000000000000000000000000000000000000000000000000000000000

a: 1152921504606846976
0001000000000000000000000000000000000000000000000000000000000000

kll@Boxy:~/terastream/acton/testzig$ 
```

With a small initial value of y it works as per above but when y = 8, which is binary 1000 and we shift that 60 bits we fall off the side of our long int and the zig cc compiled one SIGILLs (is that a verb?):

```
kll@Boxy:~/terastream/acton/testzig$ gcc test.c -o outgcc && zig cc test.c -o outzig && ./outgcc && ./outzig
a: 9223372036854775808
1000000000000000000000000000000000000000000000000000000000000000

Illegal instruction
kll@Boxy:~/terastream/acton/testzig$ 
```

gdb
```c
#0  0x00000000002017c7 in main (argc=1, argv=0x7ffc4a93af68) at test.c:21
21		y = (y >> 4) | (y << 60);
(gdb) 
```

I don't know the C standard well enough to say what should happen here. Is `zig cc` just doing something bad here? Can we shift and "trim" off the excess in some way, as to avoid the SIGILL?

**Comment by plajjan** [comment]
I asked in the zig repo: https://github.com/ziglang/zig/issues/13246

Response:
> zig cc by default sets -fsanitize-undefined, use -fno-sanitize=undefined to turn it off.

So we get similar errors with gcc if we turn that on, although it actually appears to have some fancy runtime error wrapping handler:
```
kll@Boxy:~/terastream/acton/testzig$ gcc -fsanitize=undefined z.c
z.c: In function ‘main’:
z.c:3:21: warning: left shift count >= width of type [-Wshift-count-overflow]
    3 |         long a = (1 << 100);
      |                     ^~
kll@Boxy:~/terastream/acton/testzig$ ./a.out 
z.c:3:14: runtime error: shift exponent 100 is too large for 32-bit type 'int'
```

Turning it off the zig program does not crash with SIGILL:
```
kll@Boxy:~/terastream/acton/testzig$ zig cc -fno-sanitize=undefined z.c
kll@Boxy:~/terastream/acton/testzig$ ./a.out 
```

Soooo, it seems zig is just damn careful by default, which is probably a good thing... this in turn means our code is sort of incorrect, right? I mean, we might be deliberately shifting things outside the width of our type but can we make that more explicit somehow?

**Comment by sydow** [comment]
Did you see my suggestion to change the type of y to unsigned long?
 
int main(int argc, char **argv) {
        unsigned long y = 8;
        y = (y >> 4) | (y << 60);
        long a = (long)y;

        printf("a: %lu\n", a);
        printBits(sizeof(a), &a);
        printf("\n");
        return 0;
}


With this change the warning from gcc disappears also with -fsanitize=undefined set.

It is also supported by section 6.5.7 in the C99 standard:

4   The result of E1 << E2 is E1 left-shifted E2 bit positions; vacated bits are filled with
    zeros. If E1 has an unsigned type, the value of the result is E1 × 2E2 , reduced modulo
    one more than the maximum value representable in the result type. If E1 has a signed
    type and nonnegative value, and E1 × 2E2 is representable in the result type, then that is
    the resulting value; otherwise, the behavior is undefined.

Here"2E2" should be read "2^E2". So for unsigned types the result of arbitrary left shifts are well-defined. For unsigned types, overflow leads to undefined behavior.

**Comment by plajjan** [comment]
Yes, I did see it. Thanks @sydow! Even with it fixed there are other UB leading to SIGILL, so I opted to disable this in my zig branch. It's disabled per default in current cc so this means I'm not changing anything. It's actually the saner default of zig that was changing stuff in the first place.

Anyway, I naturally want us to sanitize UB but I think it's easier to handle separately and I intend to move forward after we merge the zig branch!
